### ISSUE TITLE
Figure out how to expose TS_DEBUG_PRETENDPOINT

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#13462]
#12578 (over two commits: 42dac7c5c2598c27918528daa4405fbb79e5f0f5 and b546a6e758a9e0f7f44dd926d2cf539232426aab) added an experimental envknob `TS_DEBUG_PRETENDPOINT` for @jaxxstorm to do UDP routing experiments.  It lets you statically tell tailscaled to **pretend** that one or more **endpoints** (hence jokey portmanteau "pretendpoints") exist for the purposes of advertising to peers, but doesn't do anything to arrange for that to happen. It's assumed to be the responsibility of the user to configure their environment such that UDP packets to that `IP:port` make it to that tailscaled.

By policy, we generally don't use environment variables (or especially not debug ones like `TS_DEBUG_`) as a stable interface for configuring Tailscale. And especially not for portable features we want to support on all operating systems.  And we don't document them as a means of hopefully delaying [Hyrum's Law](https://www.hyrumslaw.com/) long enough for us to either kill the debug knob or promote it to something stable & supported. (the whole point of debug knobs and little experiments like this is to move quickly and learn something without always adding all the proper testing that'll keep it from regressing in the future, and without considering the whole design space and what things might conflict with other things... we want to continue to be able to do those ephemeral experiments to learn and not still be locked into supporting something forever when it lacks tests & proper design). In any case, if users find a debug knob and come to like it, they should have no expectation that it'll be long-time supported if it's not documented.

That said, @jaxxstorm and others are happy with the concept of what we've jokingly called "pretendpoints" (statically configuring some hypothetical endpoint the user is responsible for routing) and it's time to promote it to an actual feature.

@irbekrm added config file support for it in 57856fc0d5cbffdb81d28b0dd94e5ab2110fd58c but that's meant mostly for the Kubernetes operator, as our config file support isn't fully done yet (and certainly not GA).

The tailscaled config file mode ("declarative mode") supersedes and prevents all imperative "tailscale up" or "tailscale set" commands. It's basically all or nothing: be declarative, or be imperative.

So we only have declarative mode config support for pretendpoints.
 
We need to figure out how to expose pretendpoints to users:

* config file: done, but config file support overall isn't done enough for most users, outside Kubernetes
* `tailscale set` is possible but is profile-specific, and pretendpoints are a tailscaled daemon-wide setting, not per-profile. @nickkhyl has plans of adding tailscaled support for imperative daemon-wide settings for some Windows changes he has planned. So maybe we wait for that?
* `tailscaled --extra-endpoints=....` argument? But that precludes macOS and Windows users from setting it, without resorting back to env vars to set it on those platforms as we did for e.g. PORT, IIRC. And then making users do the kinda awkward `tailscaled-env.txt` (https://github.com/tailscale/tailscale/blob/v1.74.0/envknob/envknob.go#L493-L511) thing.
* ACL file node attributes and push it down to nodes? That's also a node-specific setting, though, equivalent to a local profile on tailscaled, so at that point you might as well make it an `ipn.Pref` and add a `tailscale set --extra-endpoints=.....`.

In any case, we need a decision.

**Comment by raggi** [comment]
I'd love to maintain consistency and obvious precedence between env vars and other ways of setting things. One possible way to do this would be to have something like `tailscale set --daemon-env TS_FOO=bar` that updates `tailscaled-env.txt`, if the update fails (say, filesystem permissions) then it could print out a patch + instructions or similar.

I like the ACL idea, but as there are some settings we don't want admins to have access to by default we'd need an opt-in mechanism which could be a `tailscaled` flag with no env equivalent. I'd love to have access to this for all prefs really, it'd be super helpful for some testing work.

**Comment by raggi** [comment]
We have a request as well to allow this to be set in a way that either sets the total set of endpoints, or with an additional setting to disallow advertising of some endpoints.

**Comment by aerusso** [comment]
Just here to make an observation in support of this feature (and perhaps a slight expansion thereof): if the specified endpoint is allowed to be a DNS name (and not just an IP address), this feature effectively provides "fallback to DNS" for setups that already have functioning DNS records.  This would mitigate concerns over migrating to tailscale if the [control plane goes down](https://tailscale.com/blog/hypergrowth-isnt-always-easy).

Sorry if this isn't the right place for this kind of comment!

**Comment by nikp123** [comment]
Please leave an option to define a just an arbitrary port (i.e. w/o a destination IP as those may change with DHCP) without restricting the users that do need the statically defined IP (as discussed above). This allows me to port forward that port specifically and tailscale always picking up the correct upstream IP.

**Comment by aerusso** [comment]
@nikp123 If I understand you correctly, the `PORT=1234` environment variable should already do that.  But, maybe I'm missing some caveat?

**Comment by nikp123** [comment]
> [@nikp123](https://github.com/nikp123) If I understand you correctly, the `PORT=1234` environment variable should already do that. But, maybe I'm missing some caveat?

Personally, for some reason, my node wouldn't advertise ``public-ip:${PORT}`` but rather ``private-ip:${PORT}`` and hence was falling back to DERP which made connections very slow. It also advertised the correct address but the wrong port (5376? I think).

Adding ``TS_DEBUG_PRETENDPORT`` made it advertise the correct ``public-ip:${PORT}`` and hence my connections started being performant again.

**Comment by dionorgua** [comment]
@nikp123 I reported this some times ago here: https://github.com/tailscale/tailscale/issues/14494

See reply from tailscale dev there. 

TL;DR: public `ip:port` is announced only if tailscale thinks that NAT is hard... And for me it was random for a long time.. Slow connection? Just restart/reboot and it may work again. No? Restart router too :) 

Honestly I don't like this "works out of the box magic" at all. I mostly migrated away from tailscale to plain wireguard because of these issues But keep using it primary for LetsEncrypt certs and docker container sidecars.

**Comment by nikp123** [comment]
> [@nikp123](https://github.com/nikp123) I reported this some times ago here: [#14494](https://github.com/tailscale/tailscale/issues/14494)
> 
> See reply from tailscale dev there.
> 
> TL;DR: public `ip:port` is announced only if tailscale thinks that NAT is hard... And for me it was random for a long time.. Slow connection? Just restart/reboot and it may work again. No? Restart router too :)
> 
> Honestly I don't like this "works out of the box magic" at all. I mostly migrated away from tailscale to plain wireguard because of these issues But keep using it primary for LetsEncrypt certs and docker container sidecars.

Thanks for the feedback, for me personally Tailscale works due to the convenience of exit nodes, it's just the server stuff that becomes hard due to issues like this. I don't mind adding a port override if that means I'll get my direct connections back.
